addons/nuxeo-retention/nuxeo-retention-core/src/main/java/org/nuxeo/retention/adapters/Record.java

```diff
@@ -0,0 +1,141 @@
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Guillaume RENARD
+ */
+package org.nuxeo.retention.adapters;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.Calendar;
+import java.util.Collections;
+import java.util.List;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.IdRef;
+import org.nuxeo.ecm.core.api.versioning.VersioningService;
+import org.nuxeo.ecm.platform.audit.service.NXAuditEventsService;
+import org.nuxeo.ecm.platform.dublincore.listener.DublinCoreListener;
+import org.nuxeo.ecm.platform.ec.notification.NotificationConstants;
+import org.nuxeo.retention.RetentionConstants;
+
+/**
+ * @since 11.1
+ */
+public class Record {
+
+    private static final Logger log = LogManager.getLogger(Record.class);
+
+    protected DocumentModel document;
+
+    public Record(final DocumentModel doc) {
+        document = doc;
+    }
+
+    public DocumentModel getDocument() {
+        return document;
+    }
+
+    public RetentionRule getRule(CoreSession session) {
+        List<String> ruleIds = getRuleIds();
+        if (ruleIds == null) {
+            return null;
+        }
+        for (String ruleId : ruleIds) {
+            IdRef ruleRef = new IdRef(ruleId);
+            if (!session.exists(ruleRef)) {
+                log.trace("Rule {} does not exist", ruleRef);
+                continue;
+            }
+            DocumentModel ruleDoc = session.getDocument(ruleRef);
+            RetentionRule rule = ruleDoc.getAdapter(RetentionRule.class);
+            if (!rule.isEnabled()) {
+                log.debug("Rule {} id disabled", ruleDoc::getPathAsString);
+                continue;
+            }
+            return rule;
+        }
+        log.debug("No active rules found for {}", document::getPathAsString);
+        return null;
+
+    }
+
+    protected List<String> getRuleIds() {
+        Serializable propertyValue = document.getPropertyValue(RetentionConstants.RECORD_RULE_IDS_PROP);
+        if (propertyValue == null) {
+            return Collections.emptyList();
+        }
+        return Arrays.asList((String[]) propertyValue);
+    }
+
+    public Calendar getSavedRetainUntil() {
+        Serializable savedRetainUntil = document.getPropertyValue(RetentionConstants.RETAIN_UNTIL_PROP);
+        if (savedRetainUntil != null) {
+            return (Calendar) savedRetainUntil;
+        }
+        return null;
+    }
+
+    public boolean isRetentionIndeterminate() {
+        if (!getDocument().isUnderRetentionOrLegalHold()) {
+            return false;
+        }
+        Calendar retainUntil = getDocument().getRetainUntil();
+        if (retainUntil == null) {
+            return false;
+        }
+        return retainUntil.compareTo(CoreSession.RETAIN_UNTIL_INDETERMINATE) == 0;
+    }
+
```

Comment:
perhaps the shorted version:
```
 return getDocument().isUnderRetentionOrLegalHold()
                && CoreSession.RETAIN_UNTIL_INDETERMINATE.compareTo(getDocument().getRetainUntil()) == 0;
```

Reply:
No, we want to do the fast path `getDocument().getRetainUntil() == null` first, because `Calendar.compareTo` is much more expensive.

Reply:
And anyway `Calendar.compareTo(null)` throws a NPE.

Reply:
you are right,